****
**Instituto Federal de Educação, Ciência e Tecnologia de Minas Gerais - Campus Ouro Preto**<br>
**Especialização em Inteligência Artificial**<br>
**Disciplina: Redes Neurais e Aprendizado Profundo**<br>
**Profs.: Dr. Agnaldo Silva, Dr. Frederico Gadelha, Dra. Sílvia Almeida**<br>
**Alunos:  Fernando Fernandes, Ivanete e Marco Antônio**
****

3. A partir da análise de um processo de destilação fracionada de petróleo observou-se que determinado óleo poderia ser classificado em duas classes de pureza {C1 e C2}, mediante a medição de três grandezas {x1, x2
e x3} que representam algumas das propriedades físico-químicas do óleo. Para tanto, pretende-se utilizar um perceptron para executar a classificação automática dessas duas classes. Assim, baseadas nas informações
coletadas do processo, formou-se o conjunto de treinamento em anexo1, tomando por convenção o valor –1 para óleo pertencente à classe C1 e o valor +1 para óleo pertencente à classe C2.

Daí, pede-se:

a. Execute dois treinamentos para a rede perceptron, inicializando-se o vetor de pesos em cada treinamento com valores aleatórios entre zero e um de tal forma que os elementos do vetor de pesos iniciais não sejam os mesmos.

In [92]:
import numpy as np

class Perceptron:
    def __init__(self, num_features, learning_rate=0.01, epochs=100):
        self.num_features = num_features
        self.learning_rate = learning_rate
        self.epochs = epochs
        self.weights = np.random.rand(num_features + 1)  # initial random weights +1 for the bias term
        self.initialWeights = self.weights
        print(f'[INFO] \tRandom initial weights: {self.weights}')

    def predict(self, inputs): # activation function
        summation = np.dot(inputs, self.weights[1:]) + self.weights[0] # activation potential: u
        return 1 if summation >= 0 else -1 # Use of the bipolar step function

    def train(self, training_data, labels):
        hasError = True
        for epoch in range(self.epochs):
            #print(f'[INFO] Epoch: {epoch}')
            hasError = False
            for inputs, label in zip(training_data, labels):
                prediction = self.predict(inputs) # return of activation function: y
                if prediction != label:
                    hasError = True
                    update = self.learning_rate * (label - prediction) # eta * (dk - y)
                    self.weights[1:] += update * inputs # update weights: w <- w + eta * (dk - y) * xk
                    self.weights[0] += update # update activation limiar: tetha <- tetha + eta * (dk - y)
                    #print(f'[INFO] Weights: {self.weights}')
            if hasError == False:
                print(f'[INFO] \tConverged after: {epoch + 1} epochs.')
                break
        print(f'[INFO] \tFinal weights: {self.weights}')
        print(f'[INFO] \tTotal of epochs: {epoch + 1}')
    
    def getInitialWeights(self):
        return self.initialWeights

    def getFinalWeights(self):
        return self.weights

In [93]:
print(f'\n[INFO] ###### Perceptron Implementation #######')
print(f'\n[INFO] Loading training dataset and labels...')
file = open('tab_treinamento1.dat', 'r')
results = list()
l = list()

for line in file:
    columns = line.split()
    columns = np.array(columns, dtype=float)
    results.append(columns[:3])
    l.append(columns[-1:])
    
training_data = np.array(results)
labels = np.array(l)
print(f'\t[INFO] OK!')

#training_data = np.array([[0.6508, 0.1097, 4.0009], [-1.4492, 0.8896, 4.4005], [2.085, 0.6876, 1.2071], [0.2626, 1.1476, 7.7985], [0.6418, 1.0234, 7.0427], [0.2569, 0.673, 8.3265], [1.1155, 0.6043, 7.4446], [0.0914, 0.3399, 7.0677], [0.0121, 0.5256, 4.6316], [-0.0429, 0.466, 5.4323], [0.434, 0.687, 8.2287], [0.2735, 1.0287, 7.1934], [0.4839, 0.4851, 7.485], [0.4089, -0.1267, 5.5019], [1.4391, 0.1614, 8.5843], [-0.9115,  -0.1973, 2.1962], [0.3654, 1.0475, 7.4858], [0.2144, 0.7515, 7.1699], [0.2013, 1.0014, 6.5489], [0.6483, 0.2183, 5.8991], [-0.1147, 0.2242, 7.2435], [-0.797, 0.8795, 3.8762], [-1.0625, 0.6366, 2.4707], [0.5307, 0.1285, 5.6883], [-1.22, 0.7777, 1.7252], [0.3957, 0.1076, 5.6623], [-0.1013, 0.5989, 7.1812], [2.4482, 0.9455, 11.2095], [2.0149, 0.6192, 10.9263], [0.2012, 0.2611, 5.4631]])
#labels = np.array([-1, -1, -1, 1, 1, -1, 1, -1, 1, 1, -1, 1, -1, -1, -1, -1, 1, 1, 1, 1, -1, 1, 1, 1, 1, -1, -1, 1, -1, 1])
print(f'\n[INFO] Getting information about training dataset...') 
print(f'[INFO] Training dataset: \n{training_data}')
print(f'[INFO] Labels of training dataset: \n{labels}')


[INFO] ###### Perceptron Implementation #######

[INFO] Loading training dataset and labels...
	[INFO] OK!

[INFO] Getting information about training dataset...
[INFO] Training dataset: 
[[-0.6508  0.1097  4.0009]
 [-1.4492  0.8896  4.4005]
 [ 2.085   0.6876 12.071 ]
 [ 0.2626  1.1476  7.7985]
 [ 0.6418  1.0234  7.0427]
 [ 0.2569  0.673   8.3265]
 [ 1.1155  0.6043  7.4446]
 [ 0.0914  0.3399  7.0677]
 [ 0.0121  0.5256  4.6316]
 [-0.0429  0.466   5.4323]
 [ 0.434   0.687   8.2287]
 [ 0.2735  1.0287  7.1934]
 [ 0.4839  0.4851  7.485 ]
 [ 0.4089 -0.1267  5.5019]
 [ 1.4391  0.1614  8.5843]
 [-0.9115 -0.1973  2.1962]
 [ 0.3654  1.0475  7.4858]
 [ 0.2144  0.7515  7.1699]
 [ 0.2013  1.0014  6.5489]
 [ 0.6483  0.2183  5.8991]
 [-0.1147  0.2242  7.2435]
 [-0.797   0.8795  3.8762]
 [-1.0625  0.6366  2.4707]
 [ 0.5307  0.1285  5.6883]
 [-1.22    0.7777  1.7252]
 [ 0.3957  0.1076  5.6623]
 [-0.1013  0.5989  7.1812]
 [ 2.4482  0.9455 11.2095]
 [ 2.0149  0.6192 10.9263]
 [ 0.2012  0.2611  5.4631]]
[

In [94]:
# Creating a Perceptron
print(f'\n[INFO] Creating a Perceptron...')
number_of_epochs = 10000
perceptron1 = Perceptron(num_features=3, learning_rate=0.01, epochs=number_of_epochs)
print(f'[INFO] \tOK!')


[INFO] Creating a Perceptron...
[INFO] 	Random initial weights: [0.17736342 0.15638614 0.47744044 0.1209934 ]
[INFO] 	OK!


In [95]:
# Training the perceptron 1
print(f'\n[INFO] Getting information about training dataset...')
print(f'[INFO] \tTraining dataset size = {training_data.shape[0]}')
print(f'[INFO] \tLabels size = {labels.shape[0]}')
print(f'[INFO] \tLimit of epochs: {number_of_epochs}')
print(f'\n[INFO] Training the Perceptron 1...')
perceptron1.train(training_data, labels)

initialWeights = perceptron1.getInitialWeights()
finalWeights = perceptron1.getFinalWeights()
print(f'\t[INFO] OK!')


[INFO] Getting information about training dataset...
[INFO] 	Training dataset size = 30
[INFO] 	Labels size = 30
[INFO] 	Limit of epochs: 10000

[INFO] Training the Perceptron 1...
[INFO] 	Converged after: 333 epochs.
[INFO] 	Final weights: [ 2.93736342  1.41135414  2.43657844 -0.7023866 ]
[INFO] 	Total of epochs: 333
	[INFO] OK!


In [96]:
# Creating a new Perceptron
print(f'\n[INFO] Creating a new Perceptron...')
number_of_epochs = 10000
perceptron2 = Perceptron(num_features=3, learning_rate=0.01, epochs=number_of_epochs)
print(f'[INFO] \tOK!')


[INFO] Creating a new Perceptron...
[INFO] 	Random initial weights: [0.44467875 0.60329962 0.69505087 0.33930767]
[INFO] 	OK!


In [97]:
# Training the perceptron 2
print(f'\n[INFO] Getting information about training dataset...')
print(f'[INFO] \tTraining dataset size = {training_data.shape[0]}')
print(f'[INFO] \tLabels size = {labels.shape[0]}')
print(f'[INFO] \tLimit of epochs: {number_of_epochs}')
print(f'\n[INFO] Training the Perceptron 2...')
perceptron2.train(training_data, labels)

print(f'\t[INFO] OK!')


[INFO] Getting information about training dataset...
[INFO] 	Training dataset size = 30
[INFO] 	Labels size = 30
[INFO] 	Limit of epochs: 10000

[INFO] Training the Perceptron 2...
[INFO] 	Converged after: 386 epochs.
[INFO] 	Final weights: [ 3.06467875  1.55732162  2.47131887 -0.73088233]
[INFO] 	Total of epochs: 386
	[INFO] OK!


b. Registre os resultados dos dois treinamentos na tabela a seguir:

<table border="1">
	<tr>
		<td rowspan="2">Treinamento</td>
		<td colspan="4">Vetor de Pesos Inicial</td>
		<td colspan="4">Vetor de Pesos Final</td>
		<td rowspan="2"> Número de Épocas</td>
	</tr>
	<tr>
		<td>b</td>
		<td>w1</td>
		<td>w2</td>
		<td>w3</td>
		<td>b</td>
		<td>w1</td>
		<td>w2</td>
		<td>w3</td>
	</tr>
	<tr>
		<td>1º (T1)</td>
		<td>0.1773634</td>
		<td>0.15638614</td>
		<td>0.47744044</td>
		<td>0.1209934</td>
		<td>2.93736342</td>
		<td>1.41135414</td>
		<td>2.43657844</td>
		<td>-0.7023866</td>
		<td>333</td>
	</tr>
	<tr>
		<td>2º (T1)</td>
		<td>0.44467875</td>
		<td>0.15638614</td>
		<td>0.69505087</td>
		<td>0.33930767</td>
		<td>3.06467875</td>
		<td>1.55732162</td>
		<td>2.47131887</td>
		<td>-0.73088233</td>
		<td>386</td>
	</tr>
</table>

c. Após o treinamento do perceptron, aplique-o na classificação automática de novas amostras de óleo (ver arquivo tab_teste1.dat), indicando-se na tabela seguinte os resultados das saídas (Classes) referentes aos dois processos de treinamento realizados no item a.

In [98]:
# Testing dataset
# test_data = np.array([[0.6508, 0.1097, 4.0009], [-1.4492, 0.8896, 4.4005], [2.085, 0.6876, 1.2071], [0.2626, 1.1476, 7.7985], [0.6418, 1.0234, 7.0427], [0.2569, 0.673, 8.3265], [1.1155, 0.6043, 7.4446], [0.0914, 0.3399, 7.0677], [0.0121, 0.5256, 4.6316], [-0.0429, 0.466, 5.4323], [0.434, 0.687, 8.2287], [0.2735, 1.0287, 7.1934], [0.4839, 0.4851, 7.485], [0.4089, -0.1267, 5.5019], [1.4391, 0.1614, 8.5843], [-0.9115,  -0.1973, 2.1962], [0.3654, 1.0475, 7.4858], [0.2144, 0.7515, 7.1699], [0.2013, 1.0014, 6.5489], [0.6483, 0.2183, 5.8991], [-0.1147, 0.2242, 7.2435], [-0.797, 0.8795, 3.8762], [-1.0625, 0.6366, 2.4707], [0.5307, 0.1285, 5.6883], [-1.22, 0.7777, 1.7252], [0.3957, 0.1076, 5.6623], [-0.1013, 0.5989, 7.1812], [2.4482, 0.9455, 11.2095], [2.0149, 0.6192, 10.9263], [0.2012, 0.2611, 5.4631]])

# Testing the perceptron
print(f'\n[INFO] Loading testing dataset...')
file = open('tab_teste1.dat', 'r')
results = list()

for line in file:
    columns = line.split()
    columns = np.array(columns, dtype=float)
    results.append(columns[:])
    
testing_data = np.array(results)
print(f'\t[INFO] OK!')

print(f'\n[INFO] Getting information about testing dataset...') 
print(f'[INFO] Testing dataset: \n{testing_data}')   
# print(f'[INFO] \tTesting dataset size = {training_data.shape[0]}')
print(f'[INFO] \tTesting dataset size = {testing_data.shape[0]}')


[INFO] Loading testing dataset...
	[INFO] OK!

[INFO] Getting information about testing dataset...
[INFO] Testing dataset: 
[[-0.3565  0.062   5.9891]
 [-0.7842  1.1267  5.5912]
 [ 0.3012  0.5611  5.8234]
 [ 0.7757  1.0648  8.0677]
 [ 0.157   0.8028  6.304 ]
 [-0.7014  1.0316  3.6005]
 [ 0.3748  0.1536  6.1537]
 [-0.692   0.9404  4.4058]
 [-1.397   0.7141  4.9263]
 [-1.8842 -0.2805  1.2548]]
[INFO] 	Testing dataset size = 10


In [99]:
print(f'\n[INFO] Running testing data with Perceptron 1...')
# for inputs in training_data:
for inputs in testing_data:
    result = perceptron1.predict(inputs)
    print(f"[INFO] \tInput: {inputs} -> Output: {result}")


[INFO] Running testing data with Perceptron 1...
[INFO] 	Input: [-0.3565  0.062   5.9891] -> Output: -1
[INFO] 	Input: [-0.7842  1.1267  5.5912] -> Output: 1
[INFO] 	Input: [0.3012 0.5611 5.8234] -> Output: 1
[INFO] 	Input: [0.7757 1.0648 8.0677] -> Output: 1
[INFO] 	Input: [0.157  0.8028 6.304 ] -> Output: 1
[INFO] 	Input: [-0.7014  1.0316  3.6005] -> Output: 1
[INFO] 	Input: [0.3748 0.1536 6.1537] -> Output: -1
[INFO] 	Input: [-0.692   0.9404  4.4058] -> Output: 1
[INFO] 	Input: [-1.397   0.7141  4.9263] -> Output: -1
[INFO] 	Input: [-1.8842 -0.2805  1.2548] -> Output: -1


In [100]:
print(f'\n[INFO] Running testing data with Perceptron 2...')
# for inputs in training_data:
for inputs in testing_data:
    result = perceptron2.predict(inputs)
    print(f"[INFO] \tInput: {inputs} -> Output: {result}")


[INFO] Running testing data with Perceptron 2...
[INFO] 	Input: [-0.3565  0.062   5.9891] -> Output: -1
[INFO] 	Input: [-0.7842  1.1267  5.5912] -> Output: 1
[INFO] 	Input: [0.3012 0.5611 5.8234] -> Output: 1
[INFO] 	Input: [0.7757 1.0648 8.0677] -> Output: 1
[INFO] 	Input: [0.157  0.8028 6.304 ] -> Output: 1
[INFO] 	Input: [-0.7014  1.0316  3.6005] -> Output: 1
[INFO] 	Input: [0.3748 0.1536 6.1537] -> Output: -1
[INFO] 	Input: [-0.692   0.9404  4.4058] -> Output: 1
[INFO] 	Input: [-1.397   0.7141  4.9263] -> Output: -1
[INFO] 	Input: [-1.8842 -0.2805  1.2548] -> Output: -1


<table border="1">
	<tr>
		<td>Amostra</td>
		<td>x1</td>
		<td>x2</td>
		<td>x3</td>
		<td>y (T1)</td>
		<td>y (T2)</td>        
	</tr>
	<tr>
		<td>1</td>
		<td>-0.3565</td>
		<td>0.0620</td>
		<td>5.9891</td>
		<td>-1</td>
		<td>-1</td>
	</tr>
	<tr>
		<td>1</td>
		<td>2</td>
		<td>3</td>
		<td>4</td>
		<td>5</td>
		<td>6</td>
	</tr>
	<tr>
		<td>1</td>
		<td>2</td>
		<td>3</td>
		<td>4</td>
		<td>5</td>
		<td>6</td>
	</tr>
	<tr>
		<td>1</td>
		<td>2</td>
		<td>3</td>
		<td>4</td>
		<td>5</td>
		<td>6</td>
	</tr>
	<tr>
		<td>1</td>
		<td>2</td>
		<td>3</td>
		<td>4</td>
		<td>5</td>
		<td>6</td>
	</tr>
	<tr>
		<td>1</td>
		<td>2</td>
		<td>3</td>
		<td>4</td>
		<td>5</td>
		<td>6</td>
	</tr>            
	<tr>
		<td>1</td>
		<td>2</td>
		<td>3</td>
		<td>4</td>
		<td>5</td>
		<td>6</td>
	</tr>
	<tr>
		<td>1</td>
		<td>2</td>
		<td>3</td>
		<td>4</td>
		<td>5</td>
		<td>6</td>
	</tr>
	<tr>
		<td>1</td>
		<td>2</td>
		<td>3</td>
		<td>4</td>
		<td>5</td>
		<td>6</td>
	</tr>
	<tr>
		<td>1</td>
		<td>2</td>
		<td>3</td>
		<td>4</td>
		<td>5</td>
		<td>6</td>
	</tr>                
</table>

d. Explique por que o número de épocas de treinamento varia a cada vez que se executa o treinamento do perceptron.

e. Qual é a principal limitação do perceptron quando aplicado em problemas de classificação de padrões?